src/particle.f90

```diff
@@ -0,0 +1,2187 @@
+  !+-------------------------------------------------------------------+
+  ! The end of the subroutine particle_bc                              |
+  !+-------------------------------------------------------------------+
+
+  !+-------------------------------------------------------------------+
+  !| This subroutine is to calculate the force on particles.           |
+  !+-------------------------------------------------------------------+
+  !| CHANGE RECORD                                                     |
+  !| -------------                                                     |
+  !| 15-11-2021  | Created by J. Fang                                  |
+  !| 01-04-2023  | Add lorentz force by J. Fang                        |
+  !+-------------------------------------------------------------------+
+  subroutine particle_force(ux1,uy1,uz1)
+    !
+    use decomp_2d, only : xsize
+    use var,       only : itime
+    use param,     only : re,t,itr
+    use mhd,       only : Bm,stuart
+    !
+    ! arguments
+    real(mytype),dimension(xsize(1),xsize(2),xsize(3)),intent(in) :: ux1,uy1,uz1
+    !
+    ! local data
+    integer :: psize,jpart,npart
+    type(partype),pointer :: pa
+    real(mytype) :: varc,vard,varq(3),maxforce,maxvdiff,vdiff,re_p,bmt(3)
+    !
+    logical,save :: firstcal=.true.
+    !
+    psize=msize(partpack)
+    !
+    npart=0
+    !
+    maxforce=0._mytype
+    maxvdiff=0._mytype
+    !
+    call field_var(ux1,uy1,uz1)
+    
+
+    do jpart=1,psize
+      !
+      pa=>partpack(jpart)
+      !
+      if(pa%new) cycle
+      !
+      pa%v(:) = pa%vf(:)
+      pa%f(:) = 0._mytype
+      !
+    enddo
+    !
+    !
+  end subroutine particle_force
+  !+-------------------------------------------------------------------+
+  ! The end of the subroutine particle_force                           |
+  !+-------------------------------------------------------------------+
+
+  !+-------------------------------------------------------------------+
+  !| this function is to retune the size of a array                    |
+  !+-------------------------------------------------------------------+
+  !| CHANGE RECORD                                                     |
+  !| -------------                                                     |
+  !| 16-Jun-2022  | Created by J. Fang STFC Daresbury Laboratory       |
+  !+-------------------------------------------------------------------+
+  pure function size_particle(var) result(nsize)
+    !
+    type(partype),allocatable,intent(in) :: var(:)
+    integer :: nsize
+    !
+    if(allocated(var)) then
+      nsize=size(var)
+    else
+      nsize=0
+    endif
+    !
+    return
+    !
+  end function size_particle
+  !
+  pure function size_integer(var) result(nsize)
+    !
+    integer,allocatable,intent(in) :: var(:)
+    integer :: nsize
+    !
+    if(allocated(var)) then
+      nsize=size(var)
+    else
+      nsize=0
+    endif
+    !
+    return
+    !
+  end function size_integer
```

Comment:
I think you can write a single/generic version of this using `class(*)` as the type of `var`.

Reply:
not working.
giveing errors: There is no specific function for the generic ‘msize’